### PR TITLE
Prevalidate hints

### PR DESCRIPTION
**Pull request by vinceatbluelabs**

### CODE REVIEW

File: records_mover/db/mysql/load_options.py [hunk 1]
```diff
@@ -17,29 +17,17 @@
     MySqlCharacterSet = Literal['big5', 'binary', 'latin1', 'ucs2',
                                 'utf8', 'utf8mb4', 'utf16', 'utf16le',
```
Comment (bluelabsbutler) [inline records_mover/db/mysql/load_options.py:18]: F401 'records_mover.records.RecordsHints' imported but unused

File: records_mover/records/types.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
-from typing import Dict, Optional, Union, List, Mapping, TYPE_CHECKING
+from typing import Dict, Optional, Union, List, Mapping, NamedTuple, Any, TypeVar, TYPE_CHECKING
```
Comment (bluelabsbutler) [inline records_mover/records/types.py:1]: F401 'typing.NamedTuple' imported but unused

Comment (bluelabsbutler) [inline records_mover/records/types.py:1]: F401 'typing.Any' imported but unused

Comment (bluelabsbutler) [inline records_mover/records/types.py:1]: F401 'typing.TypeVar' imported but unused

File: records_mover/records/types.py [hunk 3]
```diff
@@ -86,18 +75,100 @@ class UrlDetailsEntry(TypedDict):
...
+                                   "YYYY-MM-DD HH24:MI:SSOF", # TODO: this is listed twice - bug in spec?
+                                   "YYYY-MM-DD HH24:MI:SSOF",
+                                   "MM/DD/YY HH24:MI"]
+else:
+    HintDateTimeFormatTz = str
+
+VALID_DATETIMEFORMATTZS: List[HintDateTimeFormatTz] = ["YYYY-MM-DD HH:MI:SSOF",
+                                                       "YYYY-MM-DD HH:MI:SS",
+                                                       "YYYY-MM-DD HH24:MI:SSOF", # TODO: this is listed twice - bug in spec?
+                                                       "YYYY-MM-DD HH24:MI:SSOF",
```
Comment (bluelabsbutler) [inline records_mover/records/types.py:154]: E261 at least two spaces before inline comment

Comment (bluelabsbutler) [inline records_mover/records/types.py:154]: E501 line too long (105 > 100 characters)

Comment (bluelabsbutler) [inline records_mover/records/types.py:154]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/types.py [hunk 4]
```diff
@@ -86,18 +75,100 @@ class UrlDetailsEntry(TypedDict):
...
+                                                       "MM/DD/YY HH24:MI"]
+
+
+if TYPE_CHECKING:
+    HintDateTimeFormat = Literal["YYYY-MM-DD HH24:MI:SS",
+                                 'YYYY-MM-DD HH:MI:SS', # TODO this isn't in spec valid, but is part of a variant
+                                 "YYYY-MM-DD HH12:MI AM",
+                                 "MM/DD/YY HH24:MI"]
```
Comment (bluelabsbutler) [inline records_mover/records/types.py:162]: E261 at least two spaces before inline comment

Comment (bluelabsbutler) [inline records_mover/records/types.py:162]: E501 line too long (125 > 100 characters)

Comment (bluelabsbutler) [inline records_mover/records/types.py:162]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/types.py [hunk 5]
```diff
@@ -86,18 +75,100 @@ class UrlDetailsEntry(TypedDict):
...
+else:
+    HintDateTimeFormat = str
+
+
+VALID_DATETIMEFORMATS: List[HintDateTimeFormat] = ["YYYY-MM-DD HH24:MI:SS",
+                                                   'YYYY-MM-DD HH:MI:SS', # TODO this isn't in spec valid, but is part of a variant
+                                                   "YYYY-MM-DD HH12:MI AM",
```
Comment (bluelabsbutler) [inline records_mover/records/types.py:169]: E261 at least two spaces before inline comment

Comment (bluelabsbutler) [inline records_mover/records/types.py:169]: E501 line too long (113 > 100 characters)

Comment (bluelabsbutler) [inline records_mover/records/types.py:169]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/types.py [hunk 6]
```diff
@@ -116,5 +187,9 @@ class UrlDetailsEntry(TypedDict):
                                               'compression': HintCompression,
                                           },
```
Comment (bluelabsbutler) [inline records_mover/records/types.py:188]: E261 at least two spaces before inline comment

Comment (bluelabsbutler) [inline records_mover/records/types.py:188]: E501 line too long (131 > 100 characters)

Comment (bluelabsbutler) [inline records_mover/records/types.py:188]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/types.py [hunk 7]
```diff
@@ -86,18 +75,100 @@ class UrlDetailsEntry(TypedDict):
...
+# TODO: combine this and cli thingie
+if TYPE_CHECKING:
+    HintCompression = Literal['GZIP', 'BZIP', 'LZO', None]
+else:
     HintCompression = Optional[str]
 
+VALID_COMPRESSIONS: List[HintCompression] = ['GZIP', 'BZIP', 'LZO', None]
+
+if TYPE_CHECKING:
+    # The trick here works on Literal[True, False] but not on bool:
```
Comment (bluelabsbutler) [inline records_mover/records/types.py:112]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/types.py [hunk 8]
```diff
@@ -86,18 +75,100 @@ class UrlDetailsEntry(TypedDict):
...
+    # TODO: This None is a bug in the spec, right?
+    HintDateFormat = Literal[None, 'YYYY-MM-DD', 'MM-DD-YYYY', 'DD-MM-YYYY', 'MM/DD/YY']
+else:
+    # TODO: This None is a bug in the spec, right?
+    HintDateFormat = Optional[str]
+
+VALID_DATEFORMATS: List[HintDateFormat] = [
+    None, 'YYYY-MM-DD', 'MM-DD-YYYY', 'DD-MM-YYYY', 'MM/DD/YY'
+]
+
```
Comment (bluelabsbutler) [inline records_mover/records/types.py:134]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/types.py [hunk 9]
```diff
@@ -86,18 +75,100 @@ class UrlDetailsEntry(TypedDict):
...
+if TYPE_CHECKING:
+    HintTimeOnlyFormat = Literal["HH12:MI AM", "HH24:MI:SS"]
+else:
```
Comment (bluelabsbutler) [inline records_mover/records/types.py:137]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/validated_records_hints.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    doublequote: HintDoublequote
+    escape: HintEscape
+    encoding: HintEncoding
+    dateformat: HintDateFormat
+    timeonlyformat: HintTimeOnlyFormat
+    datetimeformattz: HintDateTimeFormatTz
+    datetimeformat: HintDateTimeFormat
+
+    @staticmethod
+    def validate(hints: RecordsHints,
```
Comment (bluelabsbutler) [inline records_mover/records/validated_records_hints.py:32]: C901 'ValidatedRecordsHints.validate' is too complex (20)

File: records_mover/records/validated_records_hints.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+                    HintEscape, HintEncoding, HintDateFormat,
+                    HintTimeOnlyFormat, HintDateTimeFormatTz,
+                    HintDateTimeFormat)
+from .types import (VALID_COMPRESSIONS, VALID_QUOTING, VALID_ESCAPE,
+                    VALID_ENCODING, VALID_DATEFORMATS, VALID_TIMEONLYFORMATS,
+                    VALID_DATETIMEFORMATTZS, VALID_DATETIMEFORMATS)
+from .hints import cant_handle_hint
+
+
+# TODO: Read up on namedtuple vs dataclass
```
Comment (bluelabsbutler) [inline records_mover/records/validated_records_hints.py:14]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/validated_records_hints.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+# TODO: Read up on autovalidating libraries
```
Comment (bluelabsbutler) [inline records_mover/records/validated_records_hints.py:15]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

File: records_mover/records/validated_records_hints.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        quoting = validate_quoting()
+        quotechar = validate_string('quotechar')
+        doublequote = validate_boolean('doublequote')
+        escape = validate_escape()
+        encoding = validate_encoding()
+        dateformat = validate_dateformat()
+        timeonlyformat = validate_timeonlyformat()
+        datetimeformattz = validate_datetimeformattz()
+        datetimeformat = validate_datetimeformat()
+        # TODO: After one create functions
```
Comment (bluelabsbutler) [inline records_mover/records/validated_records_hints.py:136]: Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker

**Comment by vinceatbluelabs** [comment]
Abandoned in favor of #55

**Pull request closed without merging**